docs/orchestrating-elastic-stack-applications/elasticsearch/users-and-roles.asciidoc

```diff
@@ -0,0 +1,154 @@
+
+=== File realm
+
+Custom users can also be created by providing the desired link:https://www.elastic.co/guide/en/elasticsearch/reference/current/file-realm.html[file realm content]
+in Kubernetes secrets, referenced in the Elasticsearch resource.
+
+[source,yaml]
+----
+apiVersion: elasticsearch.k8s.elastic.co/{eck_crd_version}
+kind: Elasticsearch
+metadata:
+  name: elasticsearch-sample
+spec:
+  version: {version}
+  auth:
+    fileRealm:
+    - secretName: my-filerealm-secret-1
+    - secretName: my-filerealm-secret-2
+  nodeSets:
+  - name: default
+    count: 1
+----
+
+Several secrets can be referenced in the Elasticsearch specification.
+ECK aggregates their content into a single secret, mounted in every Elasticsearch Pod.
+
+Referenced secrets may be composed of 2 entries:
+
+- `users`: content of the `users` file. It specifies user names and password hashes, as described in the link:https://www.elastic.co/guide/en/elasticsearch/reference/current/file-realm.html[file realm documentation].
+- `users_roles`: content of the `users_roles` file. It associates each role to a list of users, as described in the link:https://www.elastic.co/guide/en/elasticsearch/reference/current/file-realm.html[file realm documentation].
+
+If multiple users with the same name are specified in more than one secret, the last one takes precedence.
+If multiple roles are specified in more than one secret, a single entry per role will be derived from the concatenation of its corresponding users from all secrets.
+
+The following Secret specifies three users and their respective roles:
+
+[source,yaml]
+----
+kind: Secret
+apiVersion: v1
+metadata:
+  name: my-filerealm-secret
+stringData:
+  users: |-
+    rdeniro:$2a$10$BBJ/ILiyJ1eBTYoRKxkqbuDEdYECplvxnqQ47uiowE7yGqvCEgj9W
+    alpacino:$2a$10$cNwHnElYiMYZ/T3K4PvzGeJ1KbpXZp2PfoQD.gfaVdImnHOwIuBKS
+    jacknich:{PBKDF2}50000$z1CLJt0MEFjkIK5iEfgvfnA6xq7lF25uasspsTKSo5Q=$XxCVLbaKDimOdyWgLCLJiyoiWpA/XDMe/xtVgn1r5Sg=
+  users_roles: |-
+    admin:rdeniro
+    power_user:alpacino,jacknich
+    user:jacknich
+----
+
+The content of both `users` and `users_roles` can be populated using the link:https://www.elastic.co/guide/en/elasticsearch/reference/current/users-command.html[elasticsearch-users] tool.
+
+For example, invoking the tool in a Docker container:
+
+[source,sh]
+----
+# create a folder with the 2 files
+mkdir filerealm
+touch filerealm/users filerealm/users_roles
+
+# create user 'myuser' with role 'monitoring_user'
+docker run \
+    -v ${PWD}/filerealm:/usr/share/elasticsearch/config \
+    docker.elastic.co/elasticsearch/elasticsearch:{version} \
+    bin/elasticsearch-users useradd myuser -p mypassword -r monitoring_user
+
+# create a Kubernetes secret with the file realm content
+kubectl create secret generic my-file-realm-secret --from-file filerealm
```

Comment:
This bash snippet is not great, but better than nothing, and better than downloading the release and installing Java IMHO.

Reply:
👍